docker/Dockerfile

```diff
@@ -45,33 +58,37 @@ RUN \
     zlib1g-dev \
     && rm -rf /var/lib/apt/lists/*
 
-# Copy just what we need to pip install
-COPY MANIFEST.in README.rst setup.py /synapse/
-COPY synapse/__init__.py /synapse/synapse/__init__.py
-COPY synapse/python_dependencies.py /synapse/synapse/python_dependencies.py
+WORKDIR /synapse
+
+# Copy just what we need to run `poetry install`
+COPY pyproject.toml poetry.lock README.rst /synapse/
+
+# Install to the Python installation which hosts `pip`. In this case, it's the system
+# Python.
+ENV POETRY_VIRTUALENVS_IN_PROJECT=true \
```

Comment:
right. I've not investigated in practice how this works for initial stages of Dockerfiles, but what I do know is that if you set an ENV in the final stage, you'll also get that env var in the final container. See https://docs.docker.com/engine/reference/builder/#env:

> The environment variables set using `ENV` will persist when a container is run from the resulting image. 

and:

> If an environment variable is only needed during build, and not in the final image, consider setting a value for a single command instead

My instinct here, irrespective of whether it does the right thing in practice, is to avoid the ambiguity by avoiding `ENV`.

Reply:
I think I was allowing myself to be a little lazy since I was invoking poetry twice.

Now that we're able to export requirements.txt (92c9a60bd) this isn't needed.

For context: originally `poetry export` produced a requirements.txt file which omitted hashes for certain dependencies. That's a poetry bug. We were able to workaround this by adding a missing metadata dependency to Synapse (https://github.com/matrix-org/synapse/pull/12384) and tightening a signedjson dependency (https://github.com/matrix-org/python-signedjson/pull/9). We only found the workaround last week, and I hadn't gone back to make use of requirements.txt here (if it ain't broke...).